### PR TITLE
Fixed #31301 -- Fixed crash of QuerySet.bulk_create() with mixed empty and set ForeignKeys to AutoFields on Oracle.

### PR DESCRIPTION
**Pull request by HansAarneLiblik**
Added support for nullable AutoField, SmallAutoField and BigAutoField

### CODE REVIEW

File: tests/bulk_create/models.py [hunk 1]
```diff
@@ -81,6 +81,9 @@ class NullableFields(models.Model):
     positive_small_integer_field = models.PositiveSmallIntegerField(null=True, default=4)
     small_integer_field = models.SmallIntegerField(null=True, default=5)
     time_field = models.TimeField(null=True, default=timezone.now)
+    auto_field = models.AutoField(null=True, default=6)
+    small_auto_field = models.SmallAutoField(null=True, default=7)
+    big_auto_field = models.BigAutoField(null=True, default=8)
```
Comment (felixxm) [inline tests/bulk_create/models.py]: Model cannot have multiple `AutoField`. For `AutoField` you can use auto created `pk`, for `SmallAutoField` and `BigAutoField` you need to add separate models.

Comment (HansAarneLiblik) [reply tests/bulk_create/models.py]: Are the tests always run on clean database?

Comment (felixxm) [reply tests/bulk_create/models.py]: Yes.

Comment (HansAarneLiblik) [reply tests/bulk_create/models.py]: Changed direct `models.AutoField` to `models.ForeignKey to AutoField`

File: tests/bulk_create/models.py [hunk 2]
```diff
@@ -81,6 +93,9 @@ class NullableFields(models.Model):
     positive_small_integer_field = models.PositiveSmallIntegerField(null=True, default=4)
     small_integer_field = models.SmallIntegerField(null=True, default=5)
     time_field = models.TimeField(null=True, default=timezone.now)
+    auto_field = models.ForeignKey(AutoFieldModel, on_delete=models.CASCADE, null=True)
+    small_auto_field = models.ForeignKey(SmallAutoFieldModel, on_delete=models.CASCADE, null=True)
+    big_auto_field = models.ForeignKey(BigAutoFieldModel, on_delete=models.CASCADE, null=True)
```
Comment (felixxm) [inline tests/bulk_create/models.py]: These are not a proper regression tests. ticket-31301 describes an issue with mixed types, i.e. empty and set foreign keys in a single `bulk_create()`, e.g.
```diff
diff --git a/tests/bulk_create/models.py b/tests/bulk_create/models.py
index 87203ac4f3..e97d106029 100644
--- a/tests/bulk_create/models.py
+++ b/tests/bulk_create/models.py
@@ -53,6 +53,7 @@ class Pizzeria(Restaurant):
 
 class State(models.Model):
     two_letter_code = models.CharField(max_length=2, primary_key=True)
+    country = models.ForeignKey(Country, on_delete=models.CASCADE, null=True)
 
 
 class TwoFields(models.Model):
diff --git a/tests/bulk_create/tests.py b/tests/bulk_create/tests.py
index 2fd9801d35..9b20acb15e 100644
--- a/tests/bulk_create/tests.py
+++ b/tests/bulk_create/tests.py
@@ -60,6 +60,15 @@ class BulkCreateTests(TestCase):
         ])
         self.assertEqual(Country.objects.count(), 4)
 
+    @skipUnlessDBFeature('has_bulk_insert')
+    def test_empty_set_foreign_key(self):
+        country = Country.objects.create(description='test')
+        State.objects.bulk_create([
+            State(two_letter_code='NY'),
+            State(two_letter_code='IL', country=country),
+        ])
+        self.assertEqual(State.objects.count(), 2)
+
     def test_multi_table_inheritance_unsupported(self):
         expected_message = "Can't bulk create a multi-table inherited model"
         with self.assertRaisesMessage(ValueError, expected_message):
```

Comment (HansAarneLiblik) [reply tests/bulk_create/models.py]: There are mixed types - in `test_bulk_insert_nullable_fields`, the fields by default are filled and while looping over every field, they are set to None 1 at a time

File: tests/bulk_create/tests.py [hunk 3]
```diff
@@ -234,10 +234,23 @@ def test_bulk_insert_expressions(self):
 
     @skipUnlessDBFeature('has_bulk_insert')
     def test_bulk_insert_nullable_fields(self):
+        auto_field_instance, _ = AutoFieldModel.objects.get_or_create()
+        small_auto_field_instance, _ = SmallAutoFieldModel.objects.get_or_create()
+        big_auto_field_instance, _ = BigAutoFieldModel.objects.get_or_create()
+        auto_fields = {
```
Comment (HansAarneLiblik) [inline tests/bulk_create/tests.py]: Is this good for you or would you rather these values as default in model definition ?

File: tests/bulk_create/models.py (hunk 2 above)
Comment (felixxm) [reply tests/bulk_create/models.py]: Right.

**Comment by felixxm** [comment]
@HansAarneLiblik Thanks :+1: I pushed minor edits to tests.

Welcome aboard :boat:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Pull request merged**